kotlin/workflow-tracing/src/main/java/com/squareup/workflow/diagnostic/tracing/TracingDiagnosticListener.kt

```diff
@@ -0,0 +1,398 @@
+  name: String = ""
+): TracingDiagnosticListener = TracingDiagnosticListener(name) { workflowScope ->
+  TraceEncoder(workflowScope) {
+    file.sink()
+        .buffer()
+  }
+}
+
+/**
+ * A [WorkflowDiagnosticListener] that generates a trace file that can be viewed in Chrome by
+ * visiting `chrome://tracing`.
+ *
+ * @param name If non-empty, will be used to set the "process name" in the trace file. If empty,
+ * the workflow type is used for the process name.
+ * @param encoderProvider A function that returns a [TraceEncoder] that will be used to write trace
+ * events. The function gets the [CoroutineScope] that the workflow runtime is running in.
+ */
+@Suppress("FunctionName")
+fun TracingDiagnosticListener(
+  name: String = "",
+  memoryStats: MemoryStats = RuntimeMemoryStats,
+  encoderProvider: (workflowScope: CoroutineScope) -> TraceEncoder
+): TracingDiagnosticListener =
+  TracingDiagnosticListener(memoryStats = memoryStats) { workflowScope, type ->
+    val encoder = encoderProvider(workflowScope)
+    val processName = name.ifEmpty { type }
+    encoder.createLogger(
+        processName = processName,
+        threadName = "Profiling"
+    )
+  }
+
+/**
+ * A [WorkflowDiagnosticListener] that generates a trace file that can be viewed in Chrome by
+ * visiting `chrome://tracing`.
+ *
+ * @param loggerProvider A function that returns a [TraceLogger] that will be used to write trace
+ * events. The function gets the [CoroutineScope] that the workflow runtime is running in, as well
+ * as the same workflow type description passed to [WorkflowDiagnosticListener.onWorkflowStarted].
+ */
+@UseExperimental(VeryExperimentalWorkflow::class)
+class TracingDiagnosticListener(
+  private val memoryStats: MemoryStats = RuntimeMemoryStats,
+  private val loggerProvider: (
+    workflowScope: CoroutineScope,
+    workflowType: String
+  ) -> TraceLogger
+) : WorkflowDiagnosticListener {
+
+  /**
+   * [NONE] is fine here because it will get initialized by [onRuntimeStarted] and there's no
+   * race conditions.
+   */
+  private var logger: TraceLogger? = null
+  private var gcDetector: GcDetector? = null
+
+  private val workflowNamesById = mutableMapOf<Long, String>()
+  private val workerDescriptionsById = mutableMapOf<Long, String>()
+
+  override fun onRuntimeStarted(
+    workflowScope: CoroutineScope,
+    rootWorkflowType: String
+  ) {
+    logger = loggerProvider(workflowScope, rootWorkflowType)
+
+    // Log garbage collections in case they correlate with unusually long render times.
+    gcDetector = GcDetector {
+      logger?.log(
+          listOf(
+              Instant(
+                  name = "GC detected",
+                  scope = GLOBAL,
+                  category = "system",
+                  args = mapOf(
+                      "freeMemory" to memoryStats.freeMemory(),
+                      "totalMemory" to memoryStats.totalMemory()
+                  )
+              ),
+              createMemoryEvent()
```

Comment:
Shouldn't you create another `gcDetector` at this point? Seems like you'll only log the first gc.

Reply:
No, the GC detector does that itself.